### PR TITLE
DRD-1887: strict mode

### PR DESCRIPTION
**Pull request by ken-duck**
Story: https://issues.sonatype.org/browse/DRD-1887

Adding better exception handling and a strict mode which will allow us to prevent invalid version ranges from finding their way downstream in the feeds.

### CODE REVIEW

File: src/main/java/net/ossindex/version/InvalidRangeRuntimeException.java [hunk 1]
```diff
@@ -0,0 +1,17 @@
+package net.ossindex.version;
+
+/**
+ * The parser is not happy with us injection a regular exception, so make this one a RuntimeException. We will
```
Comment (ghost) [inline src/main/java/net/ossindex/version/InvalidRangeRuntimeException.java]: I don't understand `with us injection`.  And here's a random tomato: 🍅

Comment (ken-duck) [reply src/main/java/net/ossindex/version/InvalidRangeRuntimeException.java]: lol

s/injection/injecting/

File: src/main/java/net/ossindex/version/VersionFactory.java [hunk 2]
```diff
@@ -48,42 +49,61 @@
...
  * @author Ken Duck
- *
  */
 public class VersionFactory
 {
+  private static VersionFactory strictInstance;
+
   private static VersionFactory instance;
 
+  private boolean strict = false;
```
Comment (ghost) [inline src/main/java/net/ossindex/version/VersionFactory.java]: Can you make `strict` final?  That would help me understand if this is a switch can can't be reset.

File: src/main/java/net/ossindex/version/impl/NamedVersion.java [hunk 3]
```diff
@@ -26,22 +26,36 @@
...
  * @author Ken Duck
- *
  */
 public class NamedVersion
     implements IVersion
 {
+  // These characters have special meaning for semantic ranges, so should not be in named versions
+  private static final Pattern SEMANTIC_RANGE_SPECIAL_CHARS = Pattern.compile("[><=|&]");
+
+  // Maven ranges star with ( or [, so we should not allow any version wih these characters to become named versions
```
Comment (ghost) [inline src/main/java/net/ossindex/version/impl/NamedVersion.java]: typo: star

File: src/main/java/net/ossindex/version/impl/NamedVersion.java [hunk 4]
```diff
@@ -26,22 +26,36 @@
...
+  private static final Pattern SET_RANGE_SPECIAL_CHARS = Pattern.compile("^[\\(\\[]");
+
+  // We just won't let these charactersm happen in a named version because that would be madness
```
Comment (ghost) [inline src/main/java/net/ossindex/version/impl/NamedVersion.java]: typo: charactersm

File: src/main/java/net/ossindex/version/VersionFactory.java (hunk 2 above)
Comment (ken-duck) [reply src/main/java/net/ossindex/version/VersionFactory.java]: Indeed I can (and will!)

File: src/main/java/net/ossindex/version/impl/VersionListener.java [hunk 5]
```diff
@@ -50,10 +52,19 @@
 
   private static final Pattern startsWithDigitLetterOrHyphen = Pattern.compile("^[0-9a-zA-Z\\-]");
 
+  private boolean strict = false;
```
Comment (ghost) [inline src/main/java/net/ossindex/version/impl/VersionListener.java]: Can this be `final`?

**Review by ghost — CHANGES_REQUESTED** [review]
I looked things over, and while I don't completely understand the code, the changes seem reasonable.
Only question I have is a couple of spelling mistakes and whether `strict` can be final (I think it can).

**Review by ghost — APPROVED** [review]
Thanks for making the changes, the code looks good!

**Review by pralhadthapa — APPROVED** [review]
LGTM

**Pull request merged**